### PR TITLE
DAOS-11605 common: reuse act and tx memory

### PR DESCRIPTION
**Pull request by liuxuezhao**
Add tls cache to reuse ad_act and tx memory.
TX rate (tested by ad_mem_tests) improved from about 170K to 220K.

Required-githooks: true
Signed-off-by: Xuezhao Liu <xuezhao.liu@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/common/ad_tx.c [hunk 1]
```diff
@@ -8,8 +8,152 @@
...
+tls_act_get(int opc, size_t size)
+{
+	struct ad_act	*act;
+
+#if AD_TLS_CACHE_ENABLED
+	if (opc != UMEM_ACT_COPY && tls_cache.atc_act_nr > 0) {
+		act = d_list_pop_entry(&tls_cache.atc_act_list, struct ad_act, it_link);
+		act->it_act.ac_opc = opc;
+		tls_cache.atc_act_nr--;
+		return act;
```
Comment (NiuYawei) [inline src/common/ad_tx.c:42]: How do you know the cached act size is large enough? I think we need to define a size used for cache, and when requested size is larger than cache size, allocate in-the-fly?

Comment (liuxuezhao) [reply src/common/ad_tx.c:42]: for non-UMEM_ACT_COPY type action, all are with same size, right?
now it don't cache UMEM_ACT_COPY act.

Comment (NiuYawei) [reply src/common/ad_tx.c:42]: oh, I missed the (opc != UMEM_ACT_COPY) check. I think ACT_COPY are the major actions, what about cache larger act to satisfy most cases?

Comment (liuxuezhao) [reply src/common/ad_tx.c:50]: OK, I just added the reusing for ACT_COPY, with max payload size of 512 that is enough for most cases. seems show with better vos_perf number.

File: src/common/ad_tx.c [hunk 2]
```diff
@@ -8,8 +8,193 @@
...
+	struct umem_wal_tx_item	*item;
+	struct ad_act		*act;
+	int			 i;
+
+	tls_cache.atc_act_nr = 0;
+	D_INIT_LIST_HEAD(&tls_cache.atc_act_list);
+	for (i = 0; i < TLS_ACT_NUM; i++) {
+		D_ALLOC_PTR(act);
+		if (act == NULL)
+			return;
```
Comment (gnailzenh) [inline src/common/ad_tx.c:144]: error code?

**Review by gnailzenh — APPROVED** [review]

File: src/common/ad_tx.c (hunk 2 above)
Comment (liuxuezhao) [reply src/common/ad_tx.c:144]: intended to not return err code to simplify err handing. as the act pre-post is not must needed.

**Pull request merged**